loadable_apps/loadable_sample/wifiapp/binary_update.c

```diff
@@ -70,6 +70,86 @@ static volatile bool inf_flag = true;
 static int fail_cnt = 0;
 static unsigned int new_version = 20190422;
 
+static void binary_update_cb(void)
+{
+	printf(" =================================== \n");
+	printf("          callback funtion           \n");
+	printf(" =================================== \n");
+}
+static void binary_update_copy_binary(binary_update_info_t *binary_info)
+{
+	int read_fd;
+	int write_fd;
+	int ret;
+	int file_size;
+	int read_size;
+	uint32_t crc_hash = 0;
+	uint8_t buffer[BUFFER_SIZE];
+	binary_header_t header_data;
+
+	read_fd = open(binary_info->active_dev, O_RDONLY);
+	if (read_fd < 0) {
+		fail_cnt++;
+		printf("Failed to open %s: %d, errno: %d\n", binary_info->active_dev, ret, get_errno());
+		return;
+	}
+	write_fd = open(binary_info->inactive_dev, O_WRONLY);
+	if (write_fd < 0) {
+		fail_cnt++;
+		printf("Failed to open %s: %d, errno: %d\n", binary_info->inactive_dev, ret, get_errno());
+		return;
+	}
+	/* read to crc information. */
+	ret = read(read_fd, (FAR uint8_t *)&crc_hash, sizeof(uint32_t));
+	if (ret == ERROR) {
+		printf("Failed to read crc %s\n", binary_info->active_dev);
+		close(read_fd);
+		return;
+	}
+
+	/* Write to crc information. */
+	ret = write(write_fd, (FAR uint8_t *)&crc_hash, sizeof(uint32_t));
+	if (ret == ERROR) {
+		printf("Failed to write crc %s\n", binary_info->inactive_dev);
+		close(write_fd);
+		return;
+	}
+
+	/* Read the binary header. */
+	ret = read(read_fd, (FAR uint8_t *)&header_data, sizeof(binary_header_t));
+	if (ret != sizeof(binary_header_t)) {
+		printf("Failed to read header %s: %d\n", binary_info->active_dev, ret);
+		close(read_fd);
+		return;
+	}
+
+	/* Write the binary header. */
+	ret = write(write_fd, (FAR uint8_t *)&header_data, sizeof(binary_header_t));
+	if (ret != sizeof(binary_header_t)) {
+		printf("Failed to write header %s: %d\n", binary_info->inactive_dev, ret);
+		close(write_fd);
+		return;
+	}
+
+	/* Copy binary */
+	file_size = header_data.bin_size;
+	while (file_size > 0) {
+		read_size = file_size < BUFFER_SIZE ? file_size : BUFFER_SIZE;
+		ret = read(read_fd, (FAR uint8_t *)buffer, read_size);
+		if (ret != read_size) {
+			printf("Failed to read buffer %s: %d\n", binary_info->active_dev, ret);
+			close(read_fd);
+			return;
+		}
+		ret = write(write_fd, (FAR uint8_t *)buffer, read_size);
```

Comment:
Maybe it takes so much time, please notify to user, now we are in copying.

Reply:
Done